python/lsst/ts/tcpip/base_client_or_server.py

```diff
@@ -0,0 +1,478 @@
+        self._monitor_connection_task.set_result(None)
+
+        self.reader: asyncio.StreamReader | None = None
+        self.writer: asyncio.StreamWriter | None = None
+
+        # Task which is set done when client connects.
+        self.start_task: asyncio.Future = asyncio.create_task(self.start(**kwargs))
+
+        # Task which is set done when the client is closed.
+        self.done_task: asyncio.Future = asyncio.Future()
+
+    @property
+    def connected(self) -> bool:
+        """Return True if self.reader and self.writer are connected."""
+        return not (
+            self.reader is None
+            or self.writer is None
+            or self.reader.at_eof()
+            or self.writer.is_closing()
+        )
+
+    async def call_connect_callback(self) -> None:
+        """Call self.__connect_callback.
+
+        This is always safe to call. It only calls the callback function
+        if that function is not None and if the connection state has changed
+        since the last time this method was called.
+        """
+        connected = self.connected
+        self.log.debug(
+            f"call_connect_callback: {connected=}; was_connected={self._was_connected}"
+        )
+        if self._was_connected != connected:
+            self._was_connected = connected
+            if self.__connect_callback is not None:
+                try:
+                    # TODO DM-37477: simplify this to
+                    # `await self._connect_callback(self)`
+                    # once we drop support for sync connect_callback
+                    result = self.__connect_callback(self)
+                    if inspect.isawaitable(result):
+                        await result
+                except Exception:
+                    self.log.exception("connect_callback failed.")
+
+    @abc.abstractmethod
+    async def close(self) -> None:
+        """Close the client or server, making it unusable."""
+        raise NotImplementedError()
+
+    async def read(self, n: int) -> bytes:
+        """Read up to n bytes.
+
+        Parameters
+        ----------
+        n : `int`, optional
+            The number of bytes to read. If -1 then block until
+            the other end closes its writer, then return all data seen.
+
+        Raises
+        ------
+        `ConnectionError`
+            If the connection is lost before, or while, reading.
+        """
+        if not self.connected:
+            raise ConnectionError("Not connected")
+        assert self.reader is not None  # make mypy happy
+        try:
+            return await self.reader.read(n)
+        except (asyncio.IncompleteReadError, ConnectionError):
+            await self._close_client()
+            raise
+
+    async def readexactly(self, n: int) -> bytes:
+        """Read exactly n bytes.
+
+        Parameters
+        ----------
+        n : `int`, optional
+            The number of bytes to read.
```

Comment:
It is not "optional" because there is no default value.

Reply:
Good catch (found in two places). I had originally matched asyncio.StreamReader's default of -1, but realized that causes the read to block until the other end closes the connect -- not something we will want very often, if ever!